### PR TITLE
Amesos2 linsolfact

### PR DESCRIPTION
**Pull request by ndellingwood**
@trilinos/amesos2 

Addresses issue #3570
Updates LinearSolverFactory to include umfpack as a solver; ETI fixes for missed instantiations by using Tpetra macro, thanks @mhoemmen for assistance.

### CODE REVIEW

**Comment by ndellingwood** [comment]
Amesos2 now passes unit tests with configuration options from #3570:

Tested on MacOSX with gcc/6.5 and mpich.

```
s994290:amesos2 ndellin$ ctest
Test project /Users/ndellin/Research/trilinos/Trilinos/Build/Issue-register-factory/packages/amesos2
      Start  1: Amesos2_KLU2_UnitTests_MPI_2
 1/10 Test  #1: Amesos2_KLU2_UnitTests_MPI_2 .........................   Passed    0.24 sec
      Start  2: Amesos2_SuperLU_Solver_Test_MPI_4
 2/10 Test  #2: Amesos2_SuperLU_Solver_Test_MPI_4 ....................   Passed    0.80 sec
      Start  3: Amesos2_Superlu_UnitTests_MPI_2
 3/10 Test  #3: Amesos2_Superlu_UnitTests_MPI_2 ......................   Passed    0.17 sec
      Start  4: Amesos2_Umfpack_Solver_Test_MPI_4
 4/10 Test  #4: Amesos2_Umfpack_Solver_Test_MPI_4 ....................   Passed    0.75 sec
      Start  5: Amesos2_SolverFactory_UnitTests_MPI_4
 5/10 Test  #5: Amesos2_SolverFactory_UnitTests_MPI_4 ................   Passed    0.24 sec
      Start  6: Amesos2_Tpetra_MultiVector_Adapter_UnitTests_MPI_4
 6/10 Test  #6: Amesos2_Tpetra_MultiVector_Adapter_UnitTests_MPI_4 ...   Passed    0.16 sec
      Start  7: Amesos2_Tpetra_CrsMatrix_Adapter_UnitTests_MPI_4
 7/10 Test  #7: Amesos2_Tpetra_CrsMatrix_Adapter_UnitTests_MPI_4 .....   Passed    0.23 sec
      Start  8: Amesos2_Epetra_MultiVector_Adapter_UnitTests_MPI_4
 8/10 Test  #8: Amesos2_Epetra_MultiVector_Adapter_UnitTests_MPI_4 ...   Passed    0.15 sec
      Start  9: Amesos2_Epetra_RowMatrix_Adapter_UnitTests_MPI_4
 9/10 Test  #9: Amesos2_Epetra_RowMatrix_Adapter_UnitTests_MPI_4 .....   Passed    0.16 sec
      Start 10: Amesos2_CrsMatrix_Adapter_Consistency_Tests_MPI_4
10/10 Test #10: Amesos2_CrsMatrix_Adapter_Consistency_Tests_MPI_4 ....   Passed    0.21 sec

100% tests passed, 0 tests failed out of 10

Label Time Summary:
Amesos2    =  11.63 sec*proc (10 tests)

Total Test time (real) =   3.15 sec
```

**Comment by ndellingwood** [comment]
@srajama1 I think this is a good long term solution, it will help us stay consistent with Tpetra's configuration options.

**Comment by ndellingwood** [comment]
Failing tests were unrelated to the changes in this PR:

```
Ifpack2_BlockTriDiContainerUnitAndPerfTests_MPI_4
PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1

ROL_example_PDE-OPT_0ld_adv-diff-react_example_02_MPI_4
ROL_example_PDE-OPT_navier-stokes_example_01_MPI_4
```

@trilinos/framework are the ROL and the Panzer tests that timeout to be disabled?


Adding RETEST and crossing fingers this one makes it through...

**Comment by ndellingwood** [comment]
I believe the failing Ifpack2 test is to be addressed by PR #4608. The unit test does not make any use of Amesos2, does not make any sense this PR would trigger a failure.

File: packages/amesos2/src/Amesos2_Umfpack_def.hpp [hunk 1]
```diff
@@ -65,8 +65,8 @@ Umfpack<Matrix,Vector>::Umfpack(
   , colptr_()
   , is_contiguous_(true)
 {
-  data_.Symbolic = 0;
-  data_.Numeric = 0;
+  data_.Symbolic = NULL;
```
Comment (mhoemmen) [inline packages/amesos2/src/Amesos2_Umfpack_def.hpp]: In the future, prefer `nullptr` to `NULL`.  `nullptr` is guaranteed to be a pointer type, so it won't get confused with integer vs. pointer overloads.

Comment (ndellingwood) [reply packages/amesos2/src/Amesos2_Umfpack_def.hpp]: @mhoemmen will do, I pattern matched what I saw in other interfaces here and hope i can do a simple search/replace throughout Amesos2 to replace `NULL` with `nullptr` in a future cleanup PR.

Comment (mhoemmen) [reply packages/amesos2/src/Amesos2_Umfpack_def.hpp]: It sounds picky, but I'm trying to lean on people to embrace C++ >= 11 idioms in new code.  It helps remind them that the semantics of the language have been changing as well (e.g., there's more copy elision going on).

Comment (ndellingwood) [reply packages/amesos2/src/Amesos2_Umfpack_def.hpp]: @mhoemmen not picky at all, and I appreciate you spend time educating us, thanks!

**Comment by kyungjoo-kim** [comment]
@ndellingwood Yes. The PR testing is somewhat changed in the last few days. It used to test release build and it looks like now PR testing use debug build with bound check. That is what causes Ifpack2 failures and other time out failures.

@trilinos/framework Now this PR test depends on a chain of PRs. First I would like to confirm if we now use debug builds with bound check for PR testing. If so, submitting PR in an arbitrary order is not meaningful. 1) time out issues from Panzer and ROL must be addressed 2) other bound check or debug build related PRs e.g., PR #4608 goes in the next 3) rest of PRs.

**Comment by ndellingwood** [comment]
Adding WIP label until #4614 is addressed which will also allow #4608 to complete and merge.

**Comment by ndellingwood** [comment]
#4608 merged! Removing WIP.

**Comment by ndellingwood** [comment]
The failing builds above were fixed by other PRs merged into the develop branch (e.g. #4608), I incorrectly assumed the PR  auto-tester would test the PR merged into the current develop branch. Updating the PR with a merge of the develop branch.

**Comment by ndellingwood** [comment]
@srajama1 could you re-approve? The PR failed because it needed to updated with commits to the current develop branch.

**Comment by ndellingwood** [comment]
@srajama1 oop, I guess the push to update with the develop branch didn't stomp the previous approval so reapproval isn't needed.

**Comment by ndellingwood** [comment]
Lots of new Cuda failures unrelated to this PR as shown here...

https://testing-vm.sandia.gov/cdash/viewTest.php?onlyfailed&buildid=4716388

**Comment by jwillenbring** [comment]
> The failing builds above were fixed by other PRs merged into the develop branch (e.g. #4608), I incorrectly assumed the PR auto-tester would test the PR merged into the current develop branch. Updating the PR with a merge of the develop branch.

It sounds like this was resolved, but I wanted to confirm that the PR absolutely should be merged into the current develop without having to update the PR branch.

**Comment by srajama1** [comment]
@ndellingwood : See above.

**Comment by ndellingwood** [comment]
@jwillenbring I added an edit to the end of that statement on Friday - I don't think my manual merge of the VOTD develop branch back into my PR branch ended up being necessary as I had stated (though I can't go back and test); I glanced at the wrong build link on cdash following the 2nd retest, in that case there were a whole new set of 30-ish Cuda errors in Panzer I mentioned in #4614. The change in parallel level resolved the 1st set of failures; no idea what caused the 2nd set of failures (that I mistakenly thought were not resolved by the PR tester's merge of the develop branch); third try was a charm and the PR tests passed and merged successfully. Sorry for the confusion, thank you for the followup.

**Pull request merged**